File: ProgrammingProjects/cpp/common/src/molecule.cpp
#include <chemistry/molecule.h>

#include <chemistry/atomic_masses.h>

#include <cmath>
#include <fstream>
#include <stdexcept>
#include <string>
#include <utility>

namespace chemistry {

Molecule::Molecule(std::vector<Atom> atoms)
    : atoms_(std::move(atoms))
{
}

Molecule Molecule::from_file(const std::filesystem::path& path)
{
    std::ifstream file{path};
    if (!file) {
        throw std::runtime_error(
            "Cannot open geometry file: " + path.string()
        );
    }

    std::size_t atom_count{};
    if (!(file >> atom_count)) {
        throw std::runtime_error(
            "Failed to read atom count from: " + path.string()
        );
    }

    std::vector<Atom> atoms;
    atoms.reserve(atom_count);

    for (std::size_t index = 0; index < atom_count; ++index) {
        double atomic_number_value{};
        double x{};
        double y{};
        double z{};

        if (!(file >> atomic_number_value >> x >> y >> z)) {
            throw std::runtime_error(
                "Failed to read atom at index " +
                std::to_string(index)
            );
        }

        if (!std::isfinite(atomic_number_value) ||
            std::floor(atomic_number_value) != atomic_number_value) {
            throw std::runtime_error(
                "Invalid atomic number at index " +
                std::to_string(index)
            );
        }

        const auto atomic_number =
            static_cast<int>(atomic_number_value);

        if (atomic_number <= 0 ||
            static_cast<std::size_t>(atomic_number) >=
                atomic_masses.size()) {
            throw std::runtime_error(
                "Atomic number out of range at index " +
                std::to_string(index)
            );
        }

        atoms.push_back(Atom{
            atomic_number,
            atomic_masses[static_cast<std::size_t>(atomic_number)],
            Eigen::Vector3d{x, y, z},
        });
    }

    return Molecule{std::move(atoms)};
}

std::size_t Molecule::atom_count() const noexcept
{
    return atoms_.size();
}

const Atom& Molecule::atom(std::size_t index) const
{
    return atoms_.at(index);
}

const std::vector<Atom>& Molecule::atoms() const noexcept
{
    return atoms_;
}

} // namespace chemistry

File: ProgrammingProjects/python/chemistry/molecule.py
from pathlib import Path

import numpy as np

from .atomic_masses import ATOMIC_MASSES


class Atom:
    def __init__(
        self,
        atomic_number: int,
        position: np.ndarray,
    ) -> None:
        if not 1 <= atomic_number < len(ATOMIC_MASSES):
            raise ValueError(
                f"Invalid atomic number: {atomic_number}"
            )

        self.atomic_number = atomic_number
        self.atomic_mass = ATOMIC_MASSES[atomic_number]
        self.position = np.asarray(position, dtype=float)


class Molecule:
    def __init__(self, atoms: list[Atom]) -> None:
        self._atoms = tuple(atoms)

    @classmethod
    def from_file(cls, path: str | Path) -> "Molecule":
        geometry_path = Path(path)

        try:
            lines = geometry_path.read_text(
                encoding="utf-8"
            ).splitlines()
        except OSError as error:
            raise ValueError(
                f"Cannot open geometry file: {geometry_path}"
            ) from error

        if not lines:
            raise ValueError(
                f"Geometry file is empty: {geometry_path}"
            )

        try:
            atom_count = int(lines[0])
        except ValueError as error:
            raise ValueError(
                "Failed to read atom count"
            ) from error

        if len(lines) - 1 < atom_count:
            raise ValueError(
                f"Expected {atom_count} atoms, "
                f"but found {len(lines) - 1}"
            )

        atoms: list[Atom] = []

        for index, line in enumerate(
            lines[1 : atom_count + 1]
        ):
            fields = line.split()
            if len(fields) != 4:
                raise ValueError(
                    f"Invalid atom record at index {index}"
                )

            try:
                atomic_number_value = float(fields[0])
                coordinates = np.array(
                    fields[1:],
                    dtype=float,
                )
            except ValueError as error:
                raise ValueError(
                    f"Invalid atom record at index {index}"
                ) from error

            if not atomic_number_value.is_integer():
                raise ValueError(
                    f"Invalid atomic number at index {index}"
                )

            atoms.append(
                Atom(
                    int(atomic_number_value),
                    coordinates,
                )
            )

        return cls(atoms)

    @property
    def atom_count(self) -> int:
        return len(self._atoms)

    @property
    def atoms(self) -> tuple[Atom, ...]:
        return self._atoms

    def atom(self, index: int) -> Atom:
        if not 0 <= index < self.atom_count:
            raise IndexError("Atom index out of range")

        return self._atoms[index]

    @property
    def positions(self) -> np.ndarray:
        return np.array(
            [atom.position for atom in self._atoms]
        )

    @property
    def atomic_numbers(self) -> np.ndarray:
        return np.array(
            [atom.atomic_number for atom in self._atoms],
            dtype=int,
        )

    @property
    def masses(self) -> np.ndarray:
        return np.array(
            [atom.atomic_mass for atom in self._atoms]
        )
